src/lists.py

```diff
@@ -1,128 +1,156 @@
+    new_link.prev.next = new_link
+    new_link.next.prev = new_link
+
+
+def remove_link(link: Link[T]) -> None:
+    """Remove link from the list."""
+    link.prev.next = link.next
+    link.next.prev = link.prev
+
+
+class DLList(Generic[T]):
+    """
+    Wrapper around a doubly-linked list.
+
+    This is a circular doubly-linked list where we have a
+    dummy link that function as both the beginning and end
+    of the list. By having it, we remove multiple special
+    cases when we manipulate the list.
+
+    >>> x = DLList([1, 2, 3, 4])
+    >>> print(x)
+    [1, 2, 3, 4]
+    """
+
+    head: Link[T]  # Dummy head link
+
+    def __init__(self, seq: Iterable[T] = ()):
+        """Create a new circular list from a sequence."""
+        # Configure the head link.
+        # We are violating the type invariants this one place,
+        # but only here, so we ask the checker to just ignore it.
+        # Once the head element is configured we promise not to do
+        # it again.
+        self.head = Link(None, None, None)  # type: ignore
+        self.head.prev = self.head
+        self.head.next = self.head
+
+        # Add elements to the list, exploiting that self.head.prev
+        # is the last element in the list, so appending means inserting
+        # after that link.
+        for val in seq:
+            insert_after(self.head.prev, val)
+
+    def __str__(self) -> str:
+        """Get string with the elements going in the next direction."""
+        elms: list[str] = []
+        link = self.head.next
+        while link and link is not self.head:
+            elms.append(str(link.val))
+            link = link.next
+        return f"[{', '.join(elms)}]"
+    __repr__ = __str__  # because why not?
+
+
+# Exercises
+
+def keep(x: DLList[T], p: Callable[[T], bool]) -> None:
+    """
+    Remove all elements from x that do not satisfy the predicate p.
+
+    >>> x = DLList([1, 2, 3, 4, 5])
+    >>> keep(x, lambda a: a % 2 == 0)
+    >>> print(x)
+    [2, 4]
+    """
+
+    if x.head.next.val == None:
+        x.head = x.head.next
+        return
+    if p(x.head.next.val) == False:
+        remove_link(x.head.next)
+    else:
+        x.head = x.head.next
+    keep(x, p)
+
+# x = DLList([1, 2, 3, 4, 5])
+# keep(x, lambda a: a % 2 == 0)
+# print(x)
+
+
+def reverse(x: DLList[T]) -> None:
+    """
+    Reverse the list x.
+
+    >>> x = DLList([1, 2, 3, 4, 5])
+    >>> reverse(x)
+    >>> print(x)
+    [5, 4, 3, 2, 1]
+    """
+
+    tmp = None
+    current = x.head.next # Initialise on first element of the list
+    prevEle, prevEle.next = x.head, None
+
+    while current is not None:
+        tmp = current.next # Save next element so we can move to it in the next iteration
+        current.next = prevEle # Make current.next link to the previous element of the list
+        prevEle = current # Prepare for next link reversal
+        current = tmp # Move to next element of the list
+
+    x.head = prevEle

```

Comment:
Looks about right. It is hell to keep track of all the references, though, and an easier solution is to just flip the two references in all the links:

```python
def reverse(x: DLList[T]) -> None:
    link = x.head  # start with head, so we flip that as well
    while link != x.head:
        link.prev, link.next = link.next, link.prev
        link = link.prev  # link.prev is the old next
```